torch/csrc/distributed/c10d/TCPStore.cpp

```diff
@@ -360,12 +360,44 @@ TCPStore::TCPStore(std::string host, const TCPStoreOptions& opts)
     std::this_thread::sleep_for(std::chrono::milliseconds(distrib(gen)));
   }
 
-  client_ = detail::TCPClient::connect(addr_, opts);
-  // TCP connection established
-  C10D_DEBUG("TCP client connected to host {}:{}", addr_.host, addr_.port);
+  // Try connecting several times -- if the server listen backlog is full it may
+  // fail on the first send in validate.
+  auto deadline = std::chrono::steady_clock::now() + opts.timeout;
+  auto retry = 0;
+  do {
+    try {
+      client_ = detail::TCPClient::connect(addr_, opts);
+      // TCP connection established
+      C10D_DEBUG("TCP client connected to host {}:{}", addr_.host, addr_.port);
+
+      // client's first query for validation
+      validate();
+
+      // success
+      break;
+    } catch (const std::exception& ex) {
+      if (deadline < std::chrono::steady_clock::now()) {
+        C10D_ERROR(
+            "TCP client failed to connect/validate to host {}:{} - timed out (try={}, timeout={}ms): {}",
+            addr_.host,
+            addr_.port,
+            retry,
+            opts.timeout.count(),
+            ex.what());
+        throw;
+      }
 
-  // client's first query for validation
-  validate();
+      C10D_WARNING(
+          "TCP client failed to connect/validate to host {}:{} - retrying (try={}, timeout={}ms): {}",
+          addr_.host,
+          addr_.port,
+          retry,
+          opts.timeout.count(),
+          ex.what());
+      std::this_thread::sleep_for(kConnectRetryDelay);
```

Comment:
Since it is a gang job, randomly selecting sleep time using uniform distribution within the range would probably improve responsiveness?

Reply:
Updated to use a shared exponential backoff with jitter implementation